### DISCUSSION TITLE
How to send requests directly (without being re-captured by mitmproxy)?

### DISCUSSION THREAD

**Discussion by changsin** [mitmproxy/mitmproxy#6649] [General]
I have an addon class that captures the traffic and sends the data to my test server for analysis.
The problem is that these requests are captured by mitmproxy again. Right now, I just return immediately if the requests were sent to my test server's domain (ignoreHosts do not work in this case).

But is there a better way? I found that you can make a request by doing this:

    new_request = http.HTTPFlow.request.make(method, url, content, headers)

But how do I send it directly? I heard that you can use the replay mechanism of mitmproxy but I don't know how. Thanks,

**Comment by Prinzhorn** [comment]
> But how do I send it directly? I heard that you can use the replay mechanism of mitmproxy but I don't know how

See https://github.com/mitmproxy/mitmproxy/blob/1a02ebb89f6765d827f2fe0086dfe5960eb6e093/examples/addons/duplicate-modify-replay.py#L15

```py
from mitmproxy import ctx

# ...

new_request = http.HTTPFlow.request.make(method, url, content, headers)
ctx.master.commands.call("replay.client", [new_request])
```

However, these requests will go through all hooks etc., which is exactly the point.

If you want to send requests out-of-band use an external HTTP library (such as [httpx](https://www.python-httpx.org/async/)) and don't specify a proxy.

**Reply by changsin** [reply]
So even if I use the replay.client, it will still be going through the proxy then? Your point about not specifying proxy serves my purpose exactly. Thank you so much.

**Reply by Prinzhorn** [reply]
> So even if I use the replay.client, it will still be going through the proxy then?

I mean, technically it doesn't "go through it". But it will trigger things like `request` hook, so you can process it as if it was an actual request.
